pkg/oauthserver/authenticator/password/keystonepassword/keystonepassword.go

```diff
@@ -4,36 +4,105 @@ import (
 	"fmt"
 	"net/http"
 	"runtime/debug"
+	"encoding/json"
 
 	"github.com/golang/glog"
+	authapi "github.com/openshift/origin/pkg/oauthserver/api"
 	"github.com/rackspace/gophercloud"
 	"github.com/rackspace/gophercloud/openstack"
+	tokens3 "github.com/rackspace/gophercloud/openstack/identity/v3/tokens"
 
 	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
 	"k8s.io/apiserver/pkg/authentication/authenticator"
 	"k8s.io/apiserver/pkg/authentication/user"
-
-	authapi "github.com/openshift/origin/pkg/oauthserver/api"
-)
+	)
 
 // keystonePasswordAuthenticator uses OpenStack keystone to authenticate a user by password
 type keystonePasswordAuthenticator struct {
-	providerName   string
-	url            string
-	client         *http.Client
-	domainName     string
-	identityMapper authapi.UserIdentityMapper
+	providerName           string
+	url                    string
+	client                 *http.Client
+	domainName             string
+	identityMapper         authapi.UserIdentityMapper
+	useKeystoneIdentity    bool
 }
 
 // New creates a new password authenticator that uses OpenStack keystone to authenticate a user by password
 // A custom transport can be provided (typically to customize TLS options like trusted roots or present a client certificate).
 // If no transport is provided, http.DefaultTransport is used
-func New(providerName string, url string, transport http.RoundTripper, domainName string, identityMapper authapi.UserIdentityMapper) authenticator.Password {
+func New(providerName string, url string, transport http.RoundTripper, domainName string, identityMapper authapi.UserIdentityMapper, useKeystoneIdentity bool) authenticator.Password {
 	if transport == nil {
 		transport = http.DefaultTransport
 	}
 	client := &http.Client{Transport: transport}
-	return &keystonePasswordAuthenticator{providerName, url, client, domainName, identityMapper}
+	return &keystonePasswordAuthenticator{providerName, url, client, domainName, identityMapper, useKeystoneIdentity}
+}
+
+// Authenticate user and return his ID from Keystone
+func GetUserIDv3(client *gophercloud.ProviderClient, options gophercloud.AuthOptions) (string, error) {
```

Comment:
don't export this function if it's not needed

Reply:
will be done